### PR TITLE
[ETHEREUM-CONTRACTS] payable macro forwarder

### LINKED ISSUE #1782: [ETHEREUM-CONTRACTS] TrustedMacros: A TrustedForwarder that rules them all
## Overview
**What?**

This is a trusted forwarder with high degree of extensibility through permission-less and user-defined "macro contracts", where:

* Macro contracts encapsulates a high level ABI call wrapper, which compresses call data as a result.
  * Benefit: it is good for L2 where calldata size matters when it comes to transaction cost.
* Additionally, it makes it a much easier and consistent interface for integrating eip-712.
   *Benefit: the transaction is more readable than otherwise still having to users to read less legible eip-712 transaction with naive eip-712 integration.

An example user-defined macro request using eip-712:

image

**How?**

1. Upon requested by the TrustedMacro contract, a user-defined macro should return a array of Superfluid batch operations, which will be send to trustedForwardBatchCall by the TrustedMacro.
2. A user-defined macro is stateless but still can query other external contracts, which eliminates a class of security issues without compromising extensibility completely.

[missing image]

**Why?**

* Allow apps to compress calldata for L2s (!), without having to register a bespoke trusted forwarder.
* Allow uniformed out-of-box integration of eip-712 legible signature!

**Deadline:**

* Prototype ready: soon.
* Integration to an application: end of Jan.

**Technical Design Document (TDD):**

## Acceptance criteria
 - [x] Prototype demoed
 - [x] Test-integration with an live app
 - [ ] Satisfies entire acceptance criteria
 - [ ] Is the appetite for the spike exhausted?
 - [ ] Documentation updated
 - [ ] Presentation to stakeholders
 - [ ] Output is a TDD
 - [ ] Recommendation on next steps

## Links/Resources
* https://superfluidhq.slack.com/archives/C04V40RFJQ7/p1704895433059129
* https://github.com/superfluid-finance/protocol-monorepo/issues/68
* https://weijiekoh.github.io/eip712-signing-demo/index.html
* https://github.com/delegatable/delegatable-sol

### PR DESCRIPTION
**Pull request by d10r**
Since batch calls are payable, making the MacroForwarder payable allows the building of macros forwarding ETH.

This PR includes a test case with a new example macro offering more than one operation (create/update/delete flow), using the convention of _paramsX_ view functions for convenient parameter encoding.

### CODE REVIEW

File: packages/ethereum-contracts/contracts/utils/MacroForwarder.sol [hunk 1]
```diff
@@ -29,11 +29,12 @@ contract MacroForwarder is ForwarderBase {
...
      * @param  m      Target macro.
      * @param  params Parameters to run the macro.
+     * If value (native coins) is provided, it is forwarded.
      */
-    function runMacro(IUserDefinedMacro m, bytes calldata params) external returns (bool)
+    function runMacro(IUserDefinedMacro m, bytes calldata params) external payable returns (bool)
     {
         ISuperfluid.Operation[] memory operations = buildBatchOperations(m, params);
-        bool retVal = _forwardBatchCall(operations);
+        bool retVal = _forwardBatchCall(operations, msg.value);
```
Comment (d10r) [inline packages/ethereum-contracts/contracts/utils/MacroForwarder.sol]: Is this good enough? What if one tx has multiple invocations of this?

Comment (d10r) [reply packages/ethereum-contracts/contracts/utils/MacroForwarder.sol]: Should be. Is isolated per external function call.

File: packages/ethereum-contracts/contracts/utils/ForwarderBase.sol [hunk 2]
```diff
@@ -25,12 +25,17 @@ abstract contract ForwarderBase {
     }
 
     function _forwardBatchCall(ISuperfluid.Operation[] memory ops) internal returns (bool) {
+        return _forwardBatchCall(ops, 0);
+    }
+
+    function _forwardBatchCall(ISuperfluid.Operation[] memory ops, uint256 valueToForward) internal returns (bool) {
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/utils/ForwarderBase.sol]: let's learn from the past and use overloaded names with caution. In this case `_forwardBatchCallWithValue` is readable at use-site without having to resort to reading the library code or any editor feature.

File: packages/ethereum-contracts/contracts/interfaces/utils/IUserDefinedMacro.sol [hunk 3]
```diff
@@ -31,23 +31,16 @@ interface IUserDefinedMacro {
     function postCheck(ISuperfluid host, bytes memory params, address msgSender) external view;
 
     /*
-     * Additional to the required interface, we recommend to implement the following function:
-     * `function getParams(...) external view returns (bytes memory);`
+     * Additional to the required interface, we recommend to implement one or multiple view functions
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/interfaces/utils/IUserDefinedMacro.sol:34]: So, we avoid breaking change to user defined macro interface. That should be okay.

File: packages/ethereum-contracts/contracts/utils/ForwarderBase.sol (hunk 2 above)
Comment (d10r) [reply packages/ethereum-contracts/contracts/utils/ForwarderBase.sol]: changed

**Review by hellwolf — APPROVED** [review]

**Pull request merged**